chapters/feature-selection.qmd

```diff
@@ -0,0 +1,525 @@
+nnet_sa_res$internal %>%
+  summarize(mean = mean(RMSE), .by = c(Iter)) %>%
+  ggplot(aes(Iter, mean)) +
+  geom_line() +
+  geom_point(cex = 1 / 2) +
+  labs(x = "SA Iteration", y = "RMSE (resampled)") +
+  geom_hline(yintercept = 0.25, col = "darkgreen", lty = 2)
+```
+
+However, looks can be deceiving. The feature selection process is incredibly variable, and adding or removing an important predictor can cause severe changes in performance. @fig-selection-sa is showing the _average_ change in performance, and the relative smoothness of this curve is somewhat deceiving. 
+
+@fig-selection-sa-resamples shows what could have happened by visualizing each of the 10 searches. The results are much more dynamic and significantly vary from resample to resample. Most (but not all) searches contain one or two precipitous drops in the RMSE with many spikes of high errors (caused by a bad permutation of the solution). These cliffs and spikes do not reproducibly occur at the same iteration or in the same way. 
+
+```{r}
+#| label: fig-selection-sa-resamples
+#| echo: false
+#| out-width: 90%
+#| fig-width: 8
+#| fig-height: 6
+#| fig-cap: The individual SA runs that make up the data in @fig-selection-sa.
+load("../RData/nnet_sa_res.RData")
+
+nnet_sa_res$internal %>%
+  ggplot(aes(Iter, RMSE)) +
+  geom_line() +
+  facet_wrap(~Resample) +
+  labs(x = "SA Iteration", y = "RMSE (Assessment Set)") +
+  geom_hline(yintercept = 0.25, col = "darkgreen", lty = 2)
+```
+
+We should keep in mind that @fig-selection-sa lets us know where, on average, the search should stop, but our actual search result (that uses the entire training set) is more likely to look like what we see in @fig-selection-sa-resamples.
+
+For this particular search, the numerically best result occurred at iteration `r nnet_sa_res$optIter` and the final model contained `r length(nnet_sa_res$optVariables)` predictors where `r xfun::numbers_to_words(sum(grepl("noise", nnet_sa_res$optVariables)))` of the twenty noise columns are still contained in the model. Our resampled estimate of the RMSE, at this iteration, was `r round(mean(nnet_sa_res$internal$RMSE[nnet_sa_res$internal$Iter == nnet_sa_res$optIter]), 3)` (a perfect model would have a value of 0.25). However, the plots of the SA performance profiles indicate a range, probably above 125 iterations, where the results are equivocal.  
+
+## Filter Methods {#sec-filters}
+
+Filters can use any scoring function that quantifies how effectively the predictor predicts the outcome [@kohavi1998wrapper;@duch2006filter]. One method is to use a univariate statistic to evaluate each predictor independently. For example, the area under the ROC curve (@sec-roc) can be computed for a binary classification outcome and a set of numeric predictors. The filter could be applied using these values by pre-defining a threshold for the statistic that defines "important enough." Alternatively, the top $p^*$ predictors can be retained and used in the supervised model. 
```

Comment:
I have started to heavily lean towards suggestion that we use thresholds instead of counts. Due to the fact that it removes issues with ties. "asking for top 10 but top 9, 10, 11, 12 have same value"